### PR TITLE
Fix for serializing pytorch tensors on GPU

### LINKED ISSUE #5519: Falling back to cludpickle when serializing a GPU torch tensor
```
import ray
import torch
ray.init()
a = torch.randn(5)
b = a.cuda()
ray.put(b)
```

will print

```
2019-08-23 20:39:14,208	WARNING worker.py:436 -- WARNING: Serializing the class <class 'torch.Tensor'> failed, falling back to cloudpickle.
```

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

At the moment, Ray will fall back to cloudpickle for CUDA pytorch tensors. Instead, it should print an error and ask the user to call .cpu() on the tensor.

## What do these changes do?

This PR re-raises the exception from pyarrow that CUDA pytorch tensors cannot be serialized.

## Related issue number

Closes https://github.com/ray-project/ray/issues/5519

## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by pcmoritz** [comment]
@kiddyboots216 This is fixing your issue, let me know what you think! We should probably have a more descriptive error message here too telling people to call .cpu() on the Tensor.

**Comment by kiddyboots216** [comment]
Perfect! Thanks Phillip. I'll use this fork in our code and it will help track down the error rather than just trying to manually inspect every single tensor in the code.

**Comment by kiddyboots216** [comment]
@pcmoritz it turns out this was actually the opposite of what I wanted. Let me summarize:
- cloudpickle has correct behavior for PyTorch objects (doesn't randomly drop numpy arrays)
- pyarrow does not have correct behavior for PyTorch objects (drops requires_grad, grad_fn, .grad, all that stuff)
- before, when Ray was falling back to cloudpickle, it would have the correct behavior with PyTorch objects, i.e. if we were using CUDA
- now that Ray doesn't fall back to cloudpickle for Torch tensors, it has the incorrect behavior

In reality we want PyArrow to have the correct behavior of Cloudpickle, that way it will be fast and correct! In the status quo, Ray has entirely different behavior for the same block of code whether we use CPU or CUDA, which is bad.

File: python/ray/worker.py [hunk 1]
```diff
@@ -423,11 +423,17 @@ def _try_store_and_register(self, object_id, value):
...
-        except TypeError:
+        except TypeError as e:
             # TypeError can happen because one of the members of the object
             # may not be serializable for cloudpickle. So we need
             # these extra fallbacks here to start from the beginning.
             # Hopefully the object could have a `__reduce__` method.
+
+            # If the type error happens because a torch tensor was on the GPU
+            # and cannot be serialized due to that, we raise it to the user:
+            if "can't convert CUDA tensor to numpy" in e.args[0]:
```
Comment (robertnishihara) [inline python/ray/worker.py]: This seems like a strange place to handle it. Wouldn't it be better to just register a custom serializer for pytorch GPU tensors that raises an exception?

**Comment by pcmoritz** [comment]
@kiddyboots216 Ok, I think I understand now. Is the problem that the pyarrow serializer is dropping the gradient tape/computation graph, i.e. if you call .grad you don't get the right gradient? If yes, that's actually a deeper issue. If the gradient tape is preserved with cloudpickle (which I'm a little surprised but that's pretty cool), we can use that for everything for now if that solves your problem. It will necessarily be slower because it has to serialize a lot more stuff (i.e. the whole gradient tape).  @zhuohan123 is thinking about how to support model parallel training (i.e. differentiate through remote functions with pytorch) in a more efficient way, he might have some insight. There is also an opportunity to make cloudpickle serialization faster I think which will also help with this.

**Comment by kiddyboots216** [comment]
@pcmoritz yes, Pyarrow is dropping the gradient tape and cloudpickle preserves it. I think Daniel and I might end up making a serializer which preserves gradient tape or at least requires_grad.

**Comment by robertnishihara** [comment]
@kiddyboots216 it may be sufficient to do `ray.register_custom_serializer(torch.Tensor, use_pickle=True)`.
